eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala

```diff
@@ -72,108 +66,91 @@ class MultiPartPaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig,
-      case Some(paymentAborted) =>
-        goto(PAYMENT_ABORTED) using paymentAborted
-      case None =>
-        // Get updated local channels (will take into account the child payments that are in-flight).
-        relayer ! GetOutgoingChannels()
-        val failedPayment = d.pending(pf.id)
-        val shouldBlacklist = shouldBlacklistChannel(pf)
-        if (shouldBlacklist) {
-          log.debug(s"ignoring channel ${getFirstHopShortChannelId(failedPayment)} to ${failedPayment.routePrefix.head.nextNodeId}")
-        }
-        val ignoreChannels = if (shouldBlacklist) d.ignoreChannels + getFirstHopShortChannelId(failedPayment) else d.ignoreChannels
-        val remainingAttempts = if (shouldBlacklist && Random.nextDouble() * math.log(d.channelsCount) > 2.0) {
-          // When we have a lot of channels, many of them may end up being a bad route prefix for the destination we're
-          // trying to reach. This is a cheap error that is detected quickly (RouteNotFound), so we don't want to count
-          // it in our payment attempts to avoid failing too fast.
-          // However we don't want to test all of our channels either which would be expensive, so we only probabilistically
-          // count the failure in our payment attempts.
-          // With the log-scale used, here are the probabilities and the corresponding number of retries:
-          //  *   10 channels -> refund 13% of failures -> with 5 initial retries we will actually try 5/(1-0.13) = ~6 times
-          //  *   20 channels -> refund 32% of failures -> with 5 initial retries we will actually try 5/(1-0.32) = ~7 times
-          //  *   50 channels -> refund 50% of failures -> with 5 initial retries we will actually try 5/(1-0.50) = ~10 times
-          //  *  100 channels -> refund 56% of failures -> with 5 initial retries we will actually try 5/(1-0.56) = ~11 times
-          //  * 1000 channels -> refund 70% of failures -> with 5 initial retries we will actually try 5/(1-0.70) = ~17 times
-          // NB: this hack won't be necessary once multi-part is directly handled by the router.
-          d.remainingAttempts + 1
-        } else {
-          d.remainingAttempts
-        }
-        goto(RETRY_WITH_UPDATED_BALANCES) using d.copy(toSend = d.toSend + failedPayment.finalPayload.amount, pending = d.pending - pf.id, failures = d.failures ++ pf.failures, ignoreChannels = ignoreChannels, remainingAttempts = remainingAttempts)
-    }
+    case Event(Status.Failure(t), d: PaymentProgress) =>
+      log.warning("router error: {}", t.getMessage)
+      if (d.ignoreChannels.nonEmpty) {
+        // Channels are mostly ignored for temporary reasons, likely because they didn't have enough balance to forward
+        // the payment. When we're retrying an MPP split, it may make sense to retry those ignored channels because with
+        // a different split, they may have enough balance to forward the payment.
```

Comment:
What do you mean by `loops`? We're consuming a payment attempt so we're making progress (in the FSM sense).

We may want to be more precise and decorate with failures, it's worth digging into. But it should be rather an improvement in a future PR based on what we'd see happening IRL I think.

Reply:
Yes a loop, not an *infinite* loop. I mean retrying the same channels over and over.

Suppose there is only one route, which fails because of a `TemporaryChannelFailure` somewhere. Won't we keep retrying this route until we have `maxAttempts`?

Reply:
Got it. Yes in your example we'd be retrying that route instead of instantly failing, that's true.
But there's a possibility that one of the retries works, because a completely unrelated HTLC at the failing node may have shifted the balance in the meantime, so it could be worth it.

And note as well that the retry randomizes the path-finding algorithm, which means we may split the amounts differently (in the case where there's more than one route) so the temporary failure may not apply to the newly split amounts.

But you're right that maybe we'd need a special case here: if we found a single route (which we can test by checking `pending.isEmpty` we may want to fail directly instead of retrying. How does that sound?

Reply:
> But there's a possibility that one of the retries works, because a completely unrelated HTLC at the failing node may have shifted the balance in the meantime, so it could be worth it.

FWIW, in our toolbox we also have the `ExcludeChannel` command which bans a channel for 1 minute (by default). I don't think assuming the balance will change in our favor in the few seconds it takes for a payment to complete is very reasonable, but the randomization makes sense. Maybe the errors should actually be handled by the routing algorithm in a black box? It would be able to make subtle trade offs like this one without spoiling the payment lifecycle.

Reply:
I don't know, what I like with the current code is that `RouteCalculation` just operates on a graph and sets of ignored/additional edges/vertices, no need for business logic around HTLCs. I think it's a good componentisation to have the business logic of choosing which nodes/channels to ignore based on failures done somewhere else.

I agree though that it's potentially unlikely to have the failing channel work when retrying instantly, so in the single route case it's probably worth aborting instantaneously. But in the case where we have multiple routes and randomization, I think this "retry without ignoring channels" makes sense.

Reply:
> I don't know, what I like with the current code is that RouteCalculation just operates on a graph and sets of ignored/additional edges/vertices, no need for business logic around HTLCs. I think it's a good componentisation to have the business logic of choosing which nodes/channels to ignore based on failures done somewhere else.

Fair enough.

> I agree though that it's potentially unlikely to have the failing channel work when retrying instantly, so in the single route case it's probably worth aborting instantaneously. But in the case where we have multiple routes and randomization, I think this "retry without ignoring channels" makes sense.

Alright, I guess that's one of the things we'll need to monitor closely at the beginning.

Reply:
I'll see if I can add a metric that tracks the success rate of this mechanism, that should reveal interesting data.

Reply:
I added a metric in https://github.com/ACINQ/eclair/pull/1439/commits/60c8ba41aaf1b6e494567c79bc4048092eb07f97 to track whether this is useful or not.

I kept the retry in all cases, here is why (the sample scenario I had in mind when I added this heuristic). Imagine we have two routes to the destination: `A -> B -> D` and `A -> C -> D` and `A -> B -> D` has the capacity to send the whole payment, but in fact the balance `B -> D` is too low for that. We'll send along `A -> B -> D` and then ignore `B -> D`. If `A -> C -> D` has capacity for only half the amount, the router will return `RouteNotFound`. But if we retry without ignoring the channel `B -> D`, since the retry is randomized, it will send a smaller amount on the `A -> B -> D` route and the remainder on the `A -> C -> D` route which will work (hopefully).

Obviously this is very theoretic, we'll see in practice how this behaves. I wasn't sure what the best metric was to track that, I used a counter but if you think something else is more suited let me know.